File: utils/test-utils/src/main/java/de/sovity/edc/extension/e2e/connector/config/PortUtils.java
/*
 * Copyright (c) 2023 sovity GmbH
 *
 *  This program and the accompanying materials are made available under the
 *  terms of the Apache License, Version 2.0 which is available at
 *  https://www.apache.org/licenses/LICENSE-2.0
 *
 *  SPDX-License-Identifier: Apache-2.0
 *
 *  Contributors:
 *       sovity GmbH - init
 *
 */

package de.sovity.edc.extension.e2e.connector.config;

import lombok.AccessLevel;
import lombok.NoArgsConstructor;

import java.io.IOException;
import java.net.ServerSocket;
import java.util.Random;

import static org.eclipse.edc.util.io.Ports.MAX_TCP_PORT;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class PortUtils {

    private static final Random RANDOM = new Random();

    public static synchronized int getFreePortRange(int size) {
        // pick a random in a reasonable range
        int firstPort = RANDOM.nextInt(10_000, 50_000);

        int currentPort = firstPort;
        do {
            if (canUsePort(currentPort)) {
                currentPort++;
            } else {
                firstPort = currentPort++;
            }
        } while (currentPort <= firstPort + size);

        return firstPort;
    }

    private static boolean canUsePort(int port) {

        if (port <= 0 || port >= MAX_TCP_PORT) {
            throw new IllegalArgumentException("Lower bound must be > 0 and < " + MAX_TCP_PORT + " and be < upperBound");
        }

        try (ServerSocket serverSocket = new ServerSocket(port)) {
            serverSocket.setReuseAddress(true);
            return true;
        } catch (IOException e) {
            return false;
        }
    }
}

File: connector/ce/libs/runtime-test-lib/src/main/java/de/sovity/edc/extension/e2e/connector/config/PortUtils.kt
/*
 * Copyright sovity GmbH and/or licensed to sovity GmbH under one or
 * more contributor license agreements. You may not use this file except
 * in compliance with the "Elastic License 2.0".
 *
 * SPDX-License-Identifier: Elastic-2.0
 */
package de.sovity.edc.extension.e2e.connector.config

import org.eclipse.edc.util.io.Ports
import java.io.IOException
import java.net.ServerSocket
import java.util.Random

object PortUtils {
    private val RANDOM = Random()

    @JvmStatic
    @Synchronized
    fun getFreePortRange(size: Int): Int {
        // pick a random port in a reasonable range
        var firstPort = RANDOM.nextInt(10_000, 50_000)

        var currentPort = firstPort
        do {
            if (canUsePort(currentPort)) {
                currentPort++
            } else {
                firstPort = currentPort++
            }
        } while (currentPort <= firstPort + size)

        return firstPort
    }

    private fun canUsePort(port: Int): Boolean {
        require(port > 0 && port < Ports.MAX_TCP_PORT) { "Lower bound must be > 0 and < " + Ports.MAX_TCP_PORT + " and be < upperBound" }

        try {
            ServerSocket(port).use { serverSocket ->
                serverSocket.setReuseAddress(true)
                return true
            }
        } catch (@Suppress("SwallowedException") e: IOException) {
            return false
        }
    }
}
